boot/bootstate20.go

```diff
@@ -373,14 +466,111 @@ func (bs20 *bootState20Base) commit() error {
 	return nil
 }
 
+type bootState20Initramfs struct {
+	modeenv *Modeenv
+}
+
+func (bi20 *bootState20Initramfs) commit() error {
+	if bi20.modeenv != nil {
+		return bi20.modeenv.Write(filepath.Join(dirs.EarlyBootUbuntuData, "system-data"))
+	}
+	return nil
+}
+
+func (bs20 *bootState20Base) chooseSnapInitramfsMount(prevChoices bootStateUpdate) (
+	sn snap.PlaceInfo,
+	nextChoices bootStateUpdate,
+	err error,
+) {
+	err = bs20.loadModeenv()
+	if err != nil {
+		return nil, nil, err
+	}
+	var prevChoicesInitramfs *bootState20Initramfs
+	if prevChoices == nil {
+		prevChoicesInitramfs = &bootState20Initramfs{
+			modeenv: bs20.modeenv,
+		}
+	} else {
+		var ok bool
+		prevChoicesInitramfs, ok = prevChoices.(*bootState20Initramfs)
+		if !ok {
+			return nil, nil, fmt.Errorf("internal error: unknown bootStateUpdate type %T, expected bootState20Initramfs", prevChoices)
+		}
+	}
+
+	// first do the generic choice of which snap to use
+	first, second, err := genericEarlyBootChooseSnap(bs20, "base")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// apply the update logic to the choices modeenv
+	switch prevChoicesInitramfs.modeenv.BaseStatus {
+	case TryStatus:
+		if second != nil {
+			// then we have the first as a try snap and we are indeed trying a
+			// new base snap, so change status to TryingStatus
+			prevChoicesInitramfs.modeenv.BaseStatus = TryingStatus
+		}
+	case TryingStatus:
+		// we tried to boot a try base snap and failed, so reset
+		prevChoicesInitramfs.modeenv.BaseStatus = DefaultStatus
+	case DefaultStatus:
+		// nothing to do
+	default:
+		// log a message about invalid setting
+		logger.Noticef("invalid setting for \"base_status\" in modeenv : %q", prevChoicesInitramfs.modeenv.BaseStatus)
+	}
+
+	return first, prevChoicesInitramfs, nil
+}
+
+//
+// snapd methods
+//
+type bootState20Snapd struct {
```

Comment:
the question is not so much that we import seed, but that so far boot was only about base and kernel and not snapd

Reply:
sure, but for UC20, now the snapd snap is part of the boot process on at least install mode and first-boot run mode